### PR TITLE
feat: Add scale subresource to Alertmanger resource

### PR DESCRIPTION
**Pull request by afzal442**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Closes: #6609 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [x] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note

```

### CODE REVIEW

**Review by simonpasquier — COMMENTED** [review]
@afzal442 are you willing to continue working on this?

In addition to the current changes, you'd need to add annotations like (except that we want `replicas` instead of `shards`)

https://github.com/prometheus-operator/prometheus-operator/blob/3f36b20cfbef2df87b7c306bebe03db5d8732ab7/pkg/apis/monitoring/v1/prometheus_types.go#L815-L817

And update the status selector field:

https://github.com/prometheus-operator/prometheus-operator/blob/3f36b20cfbef2df87b7c306bebe03db5d8732ab7/pkg/prometheus/server/operator.go#L997-L1002

**Comment by afzal442** [comment]
sorry @simonpasquier I am still on it.

**Comment by ArthurSens** [comment]
Hi Afzal, it feels like something is still missing here. Have you been able to test this change manually? :)

Were you able to understand how the new fields you've added are used to scaleup/down the replicas?

**Comment by afzal442** [comment]
Thanks Arthur. Yes I understood that how those fields work under status sub
resource fields. But any further suggestions will be great.

**Comment by ArthurSens** [comment]
> Thanks Arthur. Yes I understood that how those fields work under status sub resource fields. But any further suggestions will be great.

I was hoping to trigger you to do some research by yourself a little bit, if that's fine with you :)

What you've done was add new fields to the API, but this is still not enough for the feature to work. We follow the [operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/), meaning that we read the desired state(.Spec), perform some actions (reconciliation) and update our status(.Status) to the world to see. 

The number of replicas was already present in the Spec before this PR so you didn't have to do anything there, the PR so far adds the necessary information (selectors) to the Status API as we can see in the [scale subresource documentation](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource), but I can't see anything in the reconciliation loop.

Can you read my PR's adding scale support in Prometheus and find where I've updated the reconciliation loop to update the Prometheus .Status fields?

File: pkg/alertmanager/operator.go [hunk 1]
```diff
@@ -737,18 +738,39 @@ func (c *Operator) UpdateStatus(ctx context.Context, key string) error {
...
 	}
 
+	selectorLabels := makeSelectorLabels(a.Name)
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: selectorLabels})
+	if err != nil {
+		return fmt.Errorf("failed to create selector for alertmanager scale status: %w", err)
+	}
+
+	a.Status.Selector = selector.String()
+	a.Status.Replicas = ptr.Deref(a.Spec.Replicas, 1)
```
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:748]: the field is already set by stsReporter.Update()

File: pkg/alertmanager/operator.go [hunk 2]
```diff
@@ -1696,14 +1718,18 @@ func ListOptions(name string) metav1.ListOptions {
...
+func ApplyConfigurationFromAlertmanager(a *monitoringv1.Alertmanager, updateScaleSubresource bool) *monitoringv1ac.AlertmanagerApplyConfiguration {
 	asac := monitoringv1ac.AlertmanagerStatus().
 		WithPaused(a.Status.Paused).
 		WithReplicas(a.Status.Replicas).
 		WithAvailableReplicas(a.Status.AvailableReplicas).
 		WithUpdatedReplicas(a.Status.UpdatedReplicas).
 		WithUnavailableReplicas(a.Status.UnavailableReplicas)
 
+	if updateScaleSubresource {
+		asac = asac.WithReplicas(a.Status.Replicas).WithSelector(a.Status.Selector)
```
Comment (simonpasquier) [inline pkg/alertmanager/operator.go:1730]:
```suggestion
		asac = asac.WithSelector(a.Status.Selector)
```

**Comment by afzal442** [comment]
sorry @simonpasquier any hint on why it is failing here https://github.com/prometheus-operator/prometheus-operator/actions/runs/10186430948/job/28201890145#step:5:326? and how to fix it?

**Comment by simonpasquier** [comment]
We seem to be rate-limited by github.com. But it's unrelated to your PR.

**Comment by simonpasquier** [comment]
The fix for the CI is on the main branch. Please rebase.

File: pkg/alertmanager/operator.go (hunk 1 above)
Comment (afzal442) [reply pkg/alertmanager/operator.go:748]: Thanks, I wondered what it does? I mean does it update the status subresource selector field based on replicas?

Comment (simonpasquier) [reply pkg/alertmanager/operator.go]: among other things yes.

**Review by simonpasquier — APPROVED** [review]

**Comment by ArthurSens** [comment]
nice job @afzal442!!!

**Comment by afzal442** [comment]
my goodness, I made it. thank you @ArthurSens @simonpasquier for being super supportive. 😄

**Pull request merged**